examples/adaptive_crag/application.py

```diff
@@ -0,0 +1,509 @@
+Given a user query, extract 2-5 keywords as queries for the web search. Including the topic background and main intent.
+</task>
+
+<examples>
+
+<query>
+What is Henry Feilden’s occupation?
+</query>
+<keywords>
+Henry Feilden, occupation
+</keywords>
+
+<query>
+In what city was Billy Carlson born?
+</query>
+<keywords>
+city, Billy Carlson, born
+</keywords>
+
+<query>
+What is the religion of John Gwynn?
+</query>
+<keywords>
+religion, John Gwynn
+</keywords>
+
+<query>
+What sport does Kiribati men’s national basketball team play?
+</query>
+<keywords>
+sport, Kiribati men’s national basketball team play
+</keywords>
+
+</examples>
+
+<query>
+{query}
+</query>
+"""
+
+
+def evaluator_template(query: str, document: str) -> str:
+    return f"""\
+<task>
+You are a world-class document relevance evaluator.
+Given a user query, does the following document have the exact information to answer the question? Answer True or False.
+</task>
+
+<query>
+{query}
+</query>
+
+<document>
+{document}
+</document>
+"""
+
+
+def get_user_query() -> str:
+    query = ""
+    while not query:
+        query = input("(exit, quit, q to exit) > ").strip()
+    return query
+
+
+@action(reads=["db", "chat_history"], writes=["query", "route"])
+def router(state: State, query: str, attempts: int = ATTEMPTS) -> tuple[dict[str, str], State]:
+    """
+    Route the user `query` to the appropriate place for an answer.
+    If the `query` is directly related to a table in the database, return the table name.
+    If a web search is needed, return "web_search".
+    If the `query` can be directly answered by the assistant, return "assistant".
+    If the `query` contains "exit", "quit", or "q", return "terminate".
+    If there is a `chat_history`, it will also be used as context for the routing.
+    If it fails to route after `attempts`, it will return "assistant".
+    We use `Instructor` to ensure that the router only returns valid routes.
+    """
+    if query.lower() in ["exit", "quit", "q"]:
+        route = "terminate"
+        return {"route": route}, state.update(query=query, route=route)
+    db: DBConnection = state["db"]
+    table_names = db.table_names()
+    chat_history = state["chat_history"]
+    # using this as a `response_model` to ensure the route is valid
+    routes = Literal[*table_names, "web_search", "assistant"]  # type: ignore
+    try:
+        route = ask_gemini.create(
+            messages=[
+                user_message(
+                    route_template(table_names=table_names, query=query, chat_history=chat_history)
+                )  # type: ignore
+            ],
+            response_model=routes,  # type: ignore
+            max_retries=attempts,
+        )
+    except Exception as e:
+        print(f"Error in router: {e}")
+        route = "assistant"
+    return {"route": route}, state.update(query=query, route=route)  # type: ignore
+
+
+@action(reads=["query"], writes=["lancedb_query"])
+def rewrite_query_for_lancedb(
```

Comment:
Usually at this stage you ask for a standalone question based on the chat history (see the "conversational-rag" example). Is there a reason you don't?

Reply:
Good eyes, interesting point. The chat history is used to create a message to determine the route, but not the actual query (that comes directly from the query I think). @HamzaFarhan thoughts? Might be worth a comment or a touch-up?

Reply:
I mean something like this: https://github.com/DAGWorks-Inc/hamilton/blob/3763ae8d9826a1704f03d3ad25f58381ef023833/contrib/hamilton/contrib/dagworks/conversational_rag/__init__.py#L19

Reply:
Good idea.
So incorporating the chat history when rewriting the query. The CRAG paper just mentioned a `Query Rewriter`, so I went with this basic approach. I could tweak the template to include `chat_history` if available.